catalog/app/utils/spreadsheets/spreadsheets.ts

```diff
@@ -0,0 +1,168 @@
+import * as dateFns from 'date-fns'
+import * as R from 'ramda'
+import xlsx from 'xlsx'
+
+import * as jsonSchema from 'utils/json-schema'
+import pipeThru from 'utils/pipeThru'
+
+type MetadataValue = $TSFixMe
+
+type JsonSchema = $TSFixMe
+
+export function rowsToJson(rows: MetadataValue[][]) {
+  // TODO: R.pipe(R.tail, R.length, R.max(memo)) ?
+  const maxSize = rows.reduce((memo, row) => R.max(memo, R.tail(row).length), 0)
+  return pipeThru(rows)(
+    R.map(([key, ...values]) => {
+      const nullsTail = R.repeat(null, maxSize - values.length)
+      return [key, R.concat(values, nullsTail)]
+    }),
+    R.fromPairs,
+  )
+}
+
+export function parseSpreadsheet(
+  sheet: xlsx.WorkSheet,
+  transpose: boolean,
+): Record<string, MetadataValue> {
+  const rows = xlsx.utils.sheet_to_json<MetadataValue>(sheet, {
+    header: 1,
+  })
+  return rowsToJson(transpose ? R.transpose(rows) : rows)
+}
+
+export function readSpreadsheet(file: File): Promise<xlsx.WorkSheet> {
+  return new Promise((resolve, reject) => {
+    const reader = new FileReader()
+    reader.onabort = () => {
+      reject(new Error('abort'))
+    }
+    reader.onerror = () => {
+      reject(reader.error)
+    }
+    reader.onload = () => {
+      const workbook = xlsx.read(reader.result, { type: 'array', cellDates: true })
+      resolve(workbook.Sheets[workbook.SheetNames[0]])
+    }
+    reader.readAsArrayBuffer(file)
+  })
+}
+
+export function scoreObjectDiff(obj1: {}, obj2: {}): number {
+  const keys = Object.keys(obj1)
```

Comment:
`R.intersection(Object.keys(obj1), Object.keys(obj2)).length`? (tho your solution may be more performant)
or a 1-liner maybe ? =)
`Object.keys(obj1).reduce((memo, key) => key in obj2 ? memo + 1 : memo, 0)`

Reply:
I think that `intersection` version is more expressive. It describes the algorithm exactly: find the number of intersected root keys